01.Importing all the required libraires at once.

In [2]:
# These libraries are sufficient for our Data analysis section.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

#Importing the dataset.

In [3]:
df = pd.read_csv(r'C:\Users\Nikhil Pandey\OneDrive\Desktop\PROGRAM\Projects\Heart_desease\data\raw\diabetes_raw.csv')

In [73]:
df.columns.to_list()

['race',
 'gender',
 'age',
 'admission_type_id',
 'discharge_disposition_id',
 'admission_source_id',
 'time_in_hospital',
 'num_lab_procedures',
 'num_procedures',
 'num_medications',
 'number_outpatient',
 'number_emergency',
 'number_inpatient',
 'diag_1',
 'diag_2',
 'diag_3',
 'number_diagnoses',
 'max_glu_serum',
 'A1Cresult',
 'metformin',
 'repaglinide',
 'nateglinide',
 'chlorpropamide',
 'glimepiride',
 'glipizide',
 'glyburide',
 'pioglitazone',
 'rosiglitazone',
 'acarbose',
 'miglitol',
 'insulin',
 'glyburide-metformin',
 'glipizide-metformin',
 'change',
 'diabetesMed',
 'readmitted']

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 48 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   race                      99493 non-null   object
 1   gender                    101766 non-null  object
 2   age                       101766 non-null  object
 3   weight                    3197 non-null    object
 4   admission_type_id         101766 non-null  int64 
 5   discharge_disposition_id  101766 non-null  int64 
 6   admission_source_id       101766 non-null  int64 
 7   time_in_hospital          101766 non-null  int64 
 8   payer_code                61510 non-null   object
 9   medical_specialty         51817 non-null   object
 10  num_lab_procedures        101766 non-null  int64 
 11  num_procedures            101766 non-null  int64 
 12  num_medications           101766 non-null  int64 
 13  number_outpatient         101766 non-null  int64 
 14  numb

# General Intuation:

**An aspect for the dataset, in order to get more intuative towards the domain of these specific terms which are used in the dataset. Beacause, I'm not someone who is actually from the domain of medical practiconer.**

Defintions of the terms and information: 

# Diabetes Dataset - Medication Features Dictionary

This section provides detailed information about all **medication-related features** in the diabetes readmission dataset, prepared for Machine Learning engineers.

## 1. Individual Medications

These features are typically encoded as: **"No"**, **"Steady"**, **"Up"**, or **"Down"**.

| Medication            | Class                              | Description                                                                 | Key Relationships & Impact on Readmission |
|-----------------------|------------------------------------|-----------------------------------------------------------------------------|-------------------------------------------|
| **metformin**        | Biguanide                         | First-line oral antidiabetic drug for type 2 diabetes                      | Correlated with `diabetesMed`, `A1Cresult`. Stable use often linked to lower risk. |
| **repaglinide**      | Meglitinide (short-acting)        | Stimulates insulin release, used for post-meal glucose control             | Less common; interacts with `num_medications` |
| **nateglinide**      | Meglitinide                       | Similar to repaglinide                                                     | Low frequency |
| **chlorpropamide**   | 1st Generation Sulfonylurea       | Older sulfonylurea, higher risk of hypoglycemia                            | Rare in modern data |
| **glimepiride**      | 2nd Generation Sulfonylurea       | Common insulin secretagogue                                                | Common oral agent |
| **glipizide**        | 2nd Generation Sulfonylurea       | Very commonly used                                                         | Often used in combination |
| **glyburide**        | 2nd Generation Sulfonylurea       | Common sulfonylurea                                                        | Has combination form |
| **pioglitazone**     | Thiazolidinedione (TZD)           | Insulin sensitizer, affects lipids and weight                              | Cardiovascular considerations |
| **rosiglitazone**    | Thiazolidinedione (TZD)           | Similar to pioglitazone, restricted in many regions                        | Historical CV risk concerns |
| **acarbose**         | Alpha-glucosidase inhibitor       | Delays carbohydrate absorption                                             | GI side effects common |
| **miglitol**         | Alpha-glucosidase inhibitor       | Similar to acarbose                                                        | Rare |
| **insulin**          | Injectable hormone                | Indicates more severe diabetes or poor control                             | **Very strong predictor** of readmission |

## 2. Combination Medications

| Medication                  | Description                                      | Encoding                  | Notes |
|-----------------------------|--------------------------------------------------|---------------------------|-------|
| **glyburide-metformin**    | Fixed-dose combination of glyburide + metformin | No / Steady / Up / Down   | Indicates advanced therapy |
| **glipizide-metformin**    | Fixed-dose combination of glipizide + metformin | No / Steady / Up / Down   | Advanced oral therapy |

## 3. Related Medication Context Features

| Feature             | Type     | Description                                                      | Importance |
|---------------------|----------|------------------------------------------------------------------|----------|
| **change**          | Binary   | "Ch" = diabetes medication changed during admission, "No" = no change | High - strong predictor of readmission |
| **diabetesMed**     | Binary   | "Yes" = any diabetes medication was prescribed, "No" = none    | Very high |
| **num_medications** | Integer  | Total number of medications (diabetes + others) given during stay | High - proxy for comorbidity burden |

## Modeling Notes & Recommendations

- **Feature Engineering Ideas**:
  - Create aggregates: `num_diabetes_meds`, `has_insulin`, `has_sulfonylurea`, `has_TZD`
  - Interaction features: `insulin` × `A1Cresult`, `change` × `time_in_hospital`
  - Group rare medications to reduce sparsity

- **Key Insights**:
  - Insulin use and medication changes are among the **strongest predictors** of hospital readmission.
  - Patients on multiple diabetes medications (`num_medications` high) tend to have higher readmission rates.
  - Stable metformin-only regimens are often associated with better outcomes.

- **Encoding Tips**:
  - Use one-hot encoding or target encoding for medication states.
  - Consider ordinal encoding for "No" < "Steady" < "Up"/"Down" in some models.

---

**Dataset Context**: These features come from the UCI Diabetes 130-US Hospitals dataset for years 1999–2008.

In [5]:
print(df.columns.tolist()) # Extracting titlecims

['race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']


# EDA and Data Analysis:

In [ ]:
# A general walkthrough behind the dataset:
for col in df.columns:
    print(f"\n{'='*55}")
    print(f"COLUMN : {col}")
    print(f"dtype  : {df[col].dtype}")
    print(f"missing: {df[col].isnull().sum()} | '?' values: {(df[col] == '?').sum()}")
    print(f"unique : {df[col].nunique()}")
    print(f"values : {df[col].value_counts().head(5).to_dict()}")
    print(f"{'='*55}")
    print(f"")


COLUMN : race
dtype  : object
missing: 2273 | '?' values: 0
unique : 5
values : {'Caucasian': 76099, 'AfricanAmerican': 19210, 'Hispanic': 2037, 'Other': 1506, 'Asian': 641}


COLUMN : gender
dtype  : object
missing: 0 | '?' values: 0
unique : 3
values : {'Female': 54708, 'Male': 47055, 'Unknown/Invalid': 3}


COLUMN : age
dtype  : object
missing: 0 | '?' values: 0
unique : 10
values : {'[70-80)': 26068, '[60-70)': 22483, '[50-60)': 17256, '[80-90)': 17197, '[40-50)': 9685}


COLUMN : weight
dtype  : object
missing: 98569 | '?' values: 0
unique : 9
values : {'[75-100)': 1336, '[50-75)': 897, '[100-125)': 625, '[125-150)': 145, '[25-50)': 97}


COLUMN : admission_type_id
dtype  : int64
missing: 0 | '?' values: 0
unique : 8
values : {1: 53990, 3: 18869, 2: 18480, 6: 5291, 5: 4785}


COLUMN : discharge_disposition_id
dtype  : int64
missing: 0 | '?' values: 0
unique : 26
values : {1: 60234, 3: 13954, 6: 12902, 18: 3691, 2: 2128}


COLUMN : admission_source_id
dtype  : int64
missing: 0 | '

# Exploratory Data Analysis:

Tere are some unnecessary columns in our dataset which are just random noise for the workflow. Let's get rid of them:
* weight                    → ~97% missing, can't impute this
* payer_code                → ~40% missing, low clinical value
* medical_specialty         → 50%+ missing, too vague for prediction

* These drug columns are 99%+ "No" — they add no signal to the model:
examide
citoglipton
glimepiride-pioglitazone
metformin-rosiglitazone
metformin-pioglitazone
acetohexamide
tolbutamide
tolazamide
troglitazone

In [8]:
df.drop([
    'weight',
    'payer_code',
    'medical_specialty',
    'examide',
    'citoglipton',
    'glimepiride-pioglitazone',
    'metformin-rosiglitazone',
    'metformin-pioglitazone',
    'acetohexamide',
    'tolbutamide',
    'tolazamide',
    'troglitazone'],axis=1,inplace=True)

In [9]:
df.columns

Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'glipizide', 'glyburide', 'pioglitazone',
       'rosiglitazone', 'acarbose', 'miglitol', 'insulin',
       'glyburide-metformin', 'glipizide-metformin', 'change', 'diabetesMed',
       'readmitted'],
      dtype='object')

In [ ]:
df.shape

(101766, 36)

In [54]:
df['race'].isna().sum()

np.int64(2273)

Checking is there any null values in our df.

In [72]:
#Method 01:
print(df.columns[(df.isna()).any()])

Index(['race', 'diag_1', 'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult'], dtype='object')


In [ ]:
# M02:
for col in df.columns:
    if (df[col].isnull()).any() & (df[col].isna()).any():
        print(col)

race
diag_1
diag_2
diag_3
max_glu_serum
A1Cresult


Extracting the columns with missing terms and values:
* race   2273
* diag_1  21
* diag_2                        358
* diag_3                       1423
* max_glu_serum               96420
* A1Cresult                   84748

# Univeriate anlysis:

In [ ]:
df.columns

In [ ]:
df['readmitted'].unique()

In [ ]:
df['readmitted'].plot(kind='bar')
